### PR TITLE
Generate all boilerplate for DP scream functions

### PR DESCRIPTION
**Pull request by jgfouca**
Subroutines were gathered based on #2387 

I suspect a lot of this generated code will need to be changed but it at least gives us something to start with.

Discussion points:
1) There are a number of static constants in the fortran (plev, pcnst, etc). I assume we want these to be runtime values in cxx so I treated them like function inputs.
2) The fortran files have extensive dependencies on the rest of eam which defeats the approach we used in p3 and shoc to add a file or two from eam to our build. This means we can't realistically call f90 implementations from scream for BFB unit testing without doing a full E3SM build of eam. This could also complicate the conversion since there's potentially a lot more functionality to convert beyond the dozen or so functions.
3) The iop_data_mod module has lots of data. This will probably require an additional new struct/class in Cxx.
4) What's the story with MODEL_THETA_L? Do we need to preserve this in Cxx?
5) There appear to be a number of dimensions involved (nelemd, plev, pcnst, np). Do we want to follow the same pattern of p3/shoc and have a monolithic kernel that launches a parallel_for over one of these dimensions (I assume nelemd)?
6) Which of fortran files is considered the "entry point" to DP? In other words, the equivalent of p3_main?

Made a progress table here to track things:
https://acme-climate.atlassian.net/wiki/spaces/EAMXX/pages/3736895503/2023.Q2+C+DP-SCREAM

### CODE REVIEW

File: components/eamxx/src/physics/dp/impl/dp_apply_iop_forcing_impl.hpp
Comment (tcclevenger) [inline components/eamxx/src/physics/dp/impl/dp_apply_iop_forcing_impl.hpp:16]: In F90, this function (and some others) live in the `eam/src/dynamics/` folder with calls to homme. Not sure exactly how the hooks to homme work in eamxx, but will this need to live in the `eamxx/src/dynamics` subdirectory instead of dp?

**Review by tcclevenger — COMMENTED** [review]
For 2., it also looks like there are homme/ dependencies in the `se_iop_intr_mod.F90`. Are all the eam dependencies contained in that file as well? It seems that the routines in that file acts as a sort of an interface between `eam/src/control/*iop*.F90` routines and the rest of the model, and almost exclusively called around calls to dynamics. Maybe the routines in that file make more sense in `eamxx/src/dynamics`? 

Then, all the routines from `eam/src/control/*iop*.F90` files might be able to have traditional bfb tests? But I am a bit out of my depth right now with exactly how dp works..

I can approve if would be helpful for us to start porting simpler functions now, or we can wait on @bogensch to answer some of the structure questions.

File: components/eamxx/src/physics/dp/dp_functions_f90.cpp
Comment (jgfouca) [reply components/eamxx/src/physics/dp/dp_functions_f90.cpp:178]: These are auto-generated implementations that are not guaranteed to work.

File: components/eamxx/src/physics/dp/impl/dp_apply_iop_forcing_impl.hpp
Comment (jgfouca) [reply components/eamxx/src/physics/dp/impl/dp_apply_iop_forcing_impl.hpp:16]: Yeah, I expect some of this stuff will need to be moved but I'm not worried aboutt that right now.

**Review by tcclevenger — APPROVED** [review]

**Comment by bogensch** [comment]
Thanks @jgfouca this looks good as a start; as you said will probably need to be modified as conversion takes place.  My replies to some of your questions:

4.  anything in the MODEL_THETA_L is what we want to convert since that's SCREAM's default dycore.  Anything outside of these dependencies we will not need to convert since EAMxx doesn't know about those dycores (and we don't need to preserve the ifdef MODEL_THETA_L statements).
6.  While there is no one entry point the se_iop_intr_mod.F90 serves as the main interface of DP with the rest of the model and iop_data_mod.F90 serves as the main interface with the data needed to be used by DP.

**Review by bogensch — APPROVED** [review]

**Comment by jgfouca** [comment]
Thanks, @bogensch !

**Review by tcclevenger — APPROVED** [review]

**Pull request merged**
